src/main/haskell/matching-logic/src/Kore/MatchingLogic/ProofSystem/Minimal.hs

```diff
@@ -0,0 +1,209 @@
+    deriving (Eq, Show)
+
+{-|
+  This type has constructors for each rule of the
+  proof system.
+  It is parameterized over the exact types of parts of patterns
+  to allow working with different signatures or implementations.
+  The 'term' parameter is used where the rule must be written
+  with a literal pattern.
+  'hypothesis' refers to hypotheses of a proof rule, it can be
+  instantiated with names of the hypotheses or with the actual
+  formulas giving the conclusions of those hypotheses.
+ -}
+data MLRule sort label var term hypothesis =
+   Propositional1 term term
+ | Propositional2 term term term
+ | Propositional3 term term
+ | ModusPonens hypothesis hypothesis
+ | Generalization var hypothesis
+ | VariableSubstitution
+    (SubstitutedVariable var) term (SubstitutingVariable var)
+ | ForallRule var term term
+ | Framing label Int hypothesis
+ | PropagateOr label Int term term
+     -- ^ sigma(before ..,\phi1 \/ \phi2,.. after) <->
+     --     sigma(before ..,\phi1, .. after) <-> sigma(before ..,\phi2,.. after)
+ | PropagateExists label Int var term
+     -- ^ sigma(before ..,Ex x. phi,.. after) <-> Ex x.sigma(before ..,phi,.. after)
+ | Existence var
+     -- ^ Ex x.x
+ | Singvar var term [Int] [Int]
+ deriving (Functor, Foldable, Traversable, Show)
+
+ruleTerms :: (Applicative f)
+          => (termA -> f termB)
+          -> MLRule sort label var termA hyp
+          -> f (MLRule sort label var termB hyp)
+ruleTerms f p = case p of
+   Propositional1 a b -> Propositional1 <$> f a <*> f b
+   Propositional2 a b c -> Propositional2 <$> f a <*> f b <*> f c
+   Propositional3 a b -> Propositional3 <$> f a <*> f b
+   ModusPonens h1 h2 -> pure (ModusPonens h1 h2)
+   Generalization var hyp -> pure (Generalization var hyp)
+   VariableSubstitution var1 pat var2 -> (\t -> VariableSubstitution var1 t var2) <$> f pat
+   ForallRule v a b -> ForallRule v <$> f a <*> f b
+   Framing label pos hyp -> pure (Framing label pos hyp)
+   PropagateOr label pos a b -> PropagateOr label pos <$> f a <*> f b
+   PropagateExists label pos var t -> PropagateExists label pos var <$> f t
+   Existence var -> pure (Existence var)
+   Singvar var t path1 path2 -> (\t -> Singvar var t path1 path2) <$> f t
+
+-- | The 'MLRuleSig' synonym instantiates 'MLRule' to use
+-- the sorts, labels, and patterns from the 'IsSignature' instance 'sig'
+type MLRuleSig sig var = MLRule (Sort sig) (Label sig) var (WFPattern sig var)
+
+simpleFormulaVerifier :: formula -> Either (Error MLError) ()
+simpleFormulaVerifier _ = return ()
+
+notPat :: (IsSignature sig)
+       => Sort sig -> SigPattern sig var -> SigPattern sig var
+notPat s p = Fix (AST.Not s p)
+
+andPat,orPat :: (IsSignature sig)
+             => Sort sig -> SigPattern sig var -> SigPattern sig var -> SigPattern sig var
+andPat s p1 p2 = Fix (AST.And s p1 p2)
+orPat s p1 p2 = notPat s (andPat s (notPat s p1) (notPat s p2))
+implies s p1 p2 = orPat s (notPat s p1) p2
```

Comment:
Why no signature for `implies`? And why not call it `impliesPat`?

Reply:
These are actually dead, the rest of the file uses builders from `AST.hs`